packages/core/src/engine/Autodiff.ts

```diff
@@ -920,14 +1009,31 @@ const compileNode = (
   }
 };
 
+const polyRoots = (coeffs: number[]): number[] => {
+  const n = coeffs.length;
+  // https://en.wikipedia.org/wiki/Companion_matrix
+  const companion = Matrix.zeros(n, n);
+  for (let i = 0; i + 1 < n; i++) {
+    companion.set(i + 1, i, 1);
+    companion.set(i, n - 1, -coeffs[i]);
+  }
+  companion.set(n - 1, n - 1, -coeffs[n - 1]);
+
+  const decomp = new EigenvalueDecomposition(companion);
+  return zip2(
+    decomp.realEigenvalues,
+    decomp.imaginaryEigenvalues
+  ).map(([r, i]) => (i === 0 ? r : NaN));
```

Comment:
And this is discarding imaginary roots and treating them as NaNs?

Reply:
Exactly (and these semantics are documented on the `polyRoots` function in `engine/AutodiffFunctions`). To be precise, these aren't necessarily imaginary roots, since that implies that the real component is zero. We discard any non-real root.

Reply:
Yep, would be good to note that explicitly in case it matters in the future

Reply:
Done :+1: